pkg/webhook/gateway/v1beta1/gateway_validation.go

```diff
@@ -0,0 +1,145 @@
+*/
+
+package v1beta1
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/klog/v2"
+
+	"github.com/openyurtio/openyurt/pkg/apis/raven/v1beta1"
+)
+
+// ValidateCreate implements webhook.CustomValidator so a webhook will be registered for the type.
+func (webhook *GatewayHandler) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	gw, ok := obj.(*v1beta1.Gateway)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a Gateway but got a %T", obj))
+	}
+
+	return validate(gw)
+}
+
+// ValidateUpdate implements webhook.CustomValidator so a webhook will be registered for the type.
+func (webhook *GatewayHandler) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) error {
+	newGw, ok := newObj.(*v1beta1.Gateway)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a Gateway but got a %T", newObj))
+	}
+	oldGw, ok := oldObj.(*v1beta1.Gateway)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a Gateway} but got a %T", oldObj))
+	}
+
+	if newGw.GetName() != oldGw.GetName() {
+		return apierrors.NewBadRequest(fmt.Sprintf("gateway name can not change"))
+	}
+	if err := validate(newGw); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ValidateDelete implements webhook.CustomValidator so a webhook will be registered for the type.
+func (webhook *GatewayHandler) ValidateDelete(_ context.Context, obj runtime.Object) error {
+	gw, ok := obj.(*v1beta1.Gateway)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a Gateway but got a %T", obj))
+	}
+	return validate(gw)
+}
+
+func validate(g *v1beta1.Gateway) error {
+	var errList field.ErrorList
+
+	if g.Spec.ExposeType != "" {
+		if g.Spec.ExposeType != v1beta1.ExposeTypeLoadBalancer && g.Spec.ExposeType != v1beta1.ExposeTypePublicIP {
+			fldPath := field.NewPath("spec").Child("exposeType")
+			errList = append(errList, field.Invalid(fldPath, g.Spec.ExposeType, "the 'exposeType' field  is irregularity"))
+		}
+		if g.Spec.ExposeType == v1beta1.ExposeTypeLoadBalancer || g.Spec.ExposeType == v1beta1.ExposeTypePublicIP {
+			for i, ep := range g.Spec.Endpoints {
+				if ep.UnderNAT {
+					fldPath := field.NewPath("spec").Child(fmt.Sprintf("endpoints[%d]", i)).Child("underNAT")
+					errList = append(errList, field.Invalid(fldPath, ep.UnderNAT, fmt.Sprintf("the 'underNAT' field for exposed gateway %s/%s must be false", g.Namespace, g.Name)))
+				}
+			}
+		}
+	}
+
+	if len(g.Spec.Endpoints) != 0 {
+		underNAT := g.Spec.Endpoints[0].UnderNAT
+		for i, ep := range g.Spec.Endpoints {
+			if ep.UnderNAT != underNAT {
+				fldPath := field.NewPath("spec").Child(fmt.Sprintf("endpoints[%d]", i)).Child("underNAT")
+				errList = append(errList, field.Invalid(fldPath, ep.UnderNAT, "the 'underNAT' field in endpoints must be the same"))
+			}
```

Comment:
It seems that Line94~97 is dummy code, because Line83~86 have checked that ep.UnderNAT  in all endpoints should be false.